src/python/pants/base/dependence_analysis.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from collections import defaultdict
+from dataclasses import dataclass
+from operator import itemgetter
+from typing import (
+    ClassVar,
+    DefaultDict,
+    Dict,
+    Generic,
+    Iterable,
+    List,
+    Optional,
+    Tuple,
+    Type,
+    TypeVar,
+    cast,
+)
+
+from pants.base.specs import Specs
+from pants.engine.goal import Goal
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import Sources, Target, Targets
+from pants.engine.unions import UnionMembership, union
+from pants.util.ordered_set import FrozenOrderedSet
+
+_S = TypeVar("_S", bound=Sources)
+GoalType = Type[Goal]
+
+
+@union
+@dataclass(frozen=True)
+class DependenceAnalysisRequest(Generic[_S]):
+    """A request to determine what kind of goal data dependence exists for target source fields."""
+
+    source_field_types: ClassVar[Tuple[Type[_S], ...]] = tuple()
+    targets: Targets
+
+
+@dataclass(frozen=True)
+class DependenceAnalysisResult:
+    goal: GoalType
+    accesses: Targets
+    mutates: Optional[Targets] = None
+
+
+@dataclass(frozen=True)
+class DependenceAnalysis:
+    goal_dependence_order: FrozenOrderedSet[GoalType]
+
+    def sort(self, goals: Iterable[str], goal_map: Dict[str, GoalType]) -> Iterable[str]:
+        return sorted(goals, key=self._sort_goals_key(goal_map))
+
+    def _sort_goals_key(self, goal_map: Dict[str, GoalType]):
```

Comment:
Oh, right. Wasn't sure what type was expected here, so was kind of expecting mypy to tell me, but alas, it did not, so I forgot to come back to it..